### DISCUSSION TITLE
internal events queue unexpectedly full

### DISCUSSION THREAD

**Discussion by ClementGautier** [minio/minio#18216] [Q&A]
## Disclaimer
This issue is only applicable to a very old version of the code and only relevant to the `gateway` mode which is not supported anymore. I'm not looking for a "proper" fix because it can't exists in the current context. I'm creating this issue for reference for other people that might stumble upon this bug.

## Context
We upgraded our minio to the latest version we could use given the fact that we are stuck to an old version because we use the gateway mode. We upgraded from `minio/minio:RELEASE.2021-02-14T04-01-33Z` to `minio/minio:RELEASE.2022-10-24T18-35-07Z` and we suddenly got a lot of these `internal events queue unexpectedly full` logs.

 ## Expected Behavior
Given
https://github.com/minio/minio/blob/fc6c7949727ec261cd57fbdb02fa7575d0fd8e61/cmd/event-notification.go#L329
I expected that sendEvent would not send notification in gateway mode but given the stack trace it appears `globalNotificationSys` is not `nil`.

 ## Current Behavior
```
API: SYSTEM()
Time: 14:59:59 UTC 09/05/2023
Error: internal events queue unexpectedly full (*errors.errorString)
1: net/http/server.go:2109:http.HandlerFunc.ServeHTTP()
2: minio/cmd/object-handlers.go:1982:cmd.objectAPIHandlers.PutObjectHandler()
3: minio/cmd/event-notification.go:337:cmd.sendEvent()
4: minio/cmd/event-notification.go:222:cmd.(*EventNotifier).Send()
5: minio/internal/logger/logger.go:259:logger.LogIf()
```

## Possible Solution
Maybe it have to do with the way the notification system is initialized, I see the code changed from
https://github.com/minio/minio/blob/c4e12dc846ca04c03cab9de69c828b1351631cfc/cmd/notification.go#L1288-L1300
to 
https://github.com/minio/minio/blob/fc6c7949727ec261cd57fbdb02fa7575d0fd8e61/cmd/notification.go#L1015-L1023

Is there a way to disable the notification system as we don't use it? Or a way to know why we have so much notification target?

## Steps to Reproduce (for bugs)
Deploy minio in gateway mode. Register a lot of objects inside and stress it a little.

## Regression
Yes, we didn't see this issue when we used version `minio/minio:RELEASE.2021-02-14T04-01-33Z`

## Your Environment
* Kubernetes 1.22
* Deprecated Helm chart version `8.0.14`
* Minio `minio/minio:RELEASE.2022-10-24T18-35-07Z`

**Comment by klauspost** [comment]
Obviously we do not support this, so there is nothing actionable here.

I will convert this to a discussion in case anyone from the community has an answer for you.

**Comment by ClementGautier** [comment]
As a FYI I tried to hardcode `globalNotificationSys` to `nil`, it fixed the issue BUT we lost the metrics:
```
cannot read data: cannot scrape "http://minio-encrypted-aws.minio:9000/minio/v2/metrics/cluster?target=minio-encrypted-aws.minio": Get "http://minio-encrypted-aws.minio:9000/minio/v2/metrics/cluster?target=minio-encrypted-aws.minio": net/http: timeout awaiting response headers
```

I suppose the notification system is used by the metric collector, even in gateway mode.
I'll try to simply increase the hard limit and see if it works.

**Comment by ClementGautier** [comment]
Increasing the hard limit fixed the issue:
```diff
# internal/event/targetlist.go
	// The maximum allowed number of concurrent Send() calls to all configured notifications targets
-	maxConcurrentTargetSendCalls = 20000
+	maxConcurrentTargetSendCalls = 100000
```

**Comment by ClementGautier** [comment]
In the end the issue reappeared, I'll have to dig deeper 😭

**Reply by klauspost** [reply]
Yeah, you are just postponing the issue by a factor of 5.

**Comment by ClementGautier** [comment]
After more digging I finally ended up bypassing the send notification method as it's not used by any monitoring subsystem I need (like the s3_requests metrics). So this worked for us:

```diff
# cmd/event-notification.go
func sendEvent(args eventArgs) {
- 	args.Object.Size, _ = args.Object.GetActualSize()
- 
- 	// avoid generating a notification for REPLICA creation event.
- 	if _, ok := args.ReqParams[xhttp.MinIOSourceReplicationRequest]; ok {
- 		return
- 	}
- 	// remove sensitive encryption entries in metadata.
- 	crypto.RemoveSensitiveEntries(args.Object.UserDefined)
- 	crypto.RemoveInternalEntries(args.Object.UserDefined)
- 
- 	// globalNotificationSys is not initialized in gateway mode.
- 	if globalNotificationSys == nil {
- 		return
- 	}
- 	if globalHTTPListen.NumSubscribers(args.EventName) > 0 {
- 		globalHTTPListen.Publish(args.ToEvent(false))
- 	}
- 
- 	globalEventNotifier.Send(args)
+ 	// Do nothing as we are in gateway mode and those events are not used
}
```

**Reply by mzihlmann** [reply]
thank you for this fix! just to contribute back, we also found another issue:
https://github.com/minio/minio/blob/RELEASE.2022-10-24T18-35-07Z/cmd/disk-cache.go#L745
this line must be removed, otherwise it breaks if at some point s3 is faster than disk cache - you see the likelihood of this issue is low :smiley:.
